### PR TITLE
TSL: ScriptableNode

### PR DESCRIPTION
**Pull request by sunag**
Dependencies: https://github.com/mrdoob/three.js/pull/25683

**Description**

`ScriptableNode` might seem like a redundant tool out of context, it is a JavaScript API for Nodes and the first release for CPU/GPU Nodes. `ScriptableNode` allows the serialization of scripts and maintains their connections, this opens possibilities for Node Editors who want to use TSL as a shader language.

`ScriptableNode` consider each individual code a unique Node, following this reasoning, was designed to connect with other `ScriptableNode`.

Simple Add for GPU Node
```js
const scriptableNode = js( `

	layout = {
		elements: [
			{ name: 'a', inputType: 'node' },
			{ name: 'b', inputType: 'node' }
		]
	};

	const { add } = TSL;

	function main() {

		const a = parameters.get( 'a' ) || float();
		const b = parameters.get( 'b' ) || float();

		return add( a, b );

	}

` ).scriptable();

scriptableNode.setParameter( 'a', nodeA );
scriptableNode.setParameter( 'b', nodeB );

material = new MeshBasicNodeMaterial();
material.colorNode = scriptableNode;
```

### Visual example
image

### CODE REVIEW

**Comment by LeviPesin** [comment]
I think we should put ScriptableNode, ScriptableValueNode, FunctionNode and CallNode into a separate directory -- maybe call it "code". The idea is that these nodes provide direct low-level access to the language the nodes currently run at (JS, GLSL, WGSL).

File: examples/jsm/nodes/scriptable/ScriptableNode.js [hunk 1]
```diff
@@ -0,0 +1,427 @@
...
+		const setOutput = ( id, value ) => this.setOutput( id, value );
+
+		const parameters = new Parameters( this );
+
+		const method = this.getMethod( this.codeNode );
+		const params = [ parameters, this.local, global, refresh, setOutput, THREE, Nodes, this.editor ];
+
+		this._object = method( ...params );
+
+		const layout = this._object.layout;
```
Comment (LeviPesin) [inline examples/jsm/nodes/scriptable/ScriptableNode.js:280]: Maybe we should move this to the editor?

**Comment by sunag** [comment]
I also want to rename `accessors` to `inputs` but better to do that in another PR.

File: examples/jsm/nodes/code/CodeNode.js [hunk 2]
```diff
@@ -1,4 +1,4 @@
-import Node, { addNodeClass } from './Node.js';
+import Node, { addNodeClass } from '../core/Node.js';
 import { nodeProxy } from '../shadernode/ShaderNode.js';
```
Comment (LeviPesin) [inline examples/jsm/nodes/code/CodeNode.js:2]: '../' here should be changed to '../../'.

**Comment by LeviPesin** [comment]
> I also want to rename `accessors` to `inputs` but better to do that in another PR.

Maybe `objects` will fit more? Because I think things like BufferNode, TextureNode, ReferenceNode (maybe we should move it there?) are more understandable as "objects" rather than "inputs".

**Comment by sunag** [comment]
> Maybe objects will fit more? Because I think things like BufferNode, TextureNode, ReferenceNode (maybe we should move it there?) are more understandable as "objects" rather than "inputs".

Inputs refers to `vars, uniforms, textures, buffers, etc`, while objects is a broader abstraction, but we can certainly have both and make that split.

File: examples/jsm/nodes/code/CodeNode.js (hunk 2 above)
Comment (sunag) [reply examples/jsm/nodes/code/CodeNode.js:2]: Why? doesn't seem to work that way

File: examples/jsm/nodes/code/ScriptableNode.js [hunk 3]
```diff
@@ -0,0 +1,490 @@
+import * as THREE from 'three';
+import * as Nodes from 'three/nodes';
```
Comment (LeviPesin) [inline examples/jsm/nodes/code/ScriptableNode.js:2]: This actually creates a circular reference... Maybe it's better to do something similar to such function used in NodeLoader.

File: examples/jsm/nodes/code/CodeNode.js (hunk 2 above)
Comment (LeviPesin) [reply examples/jsm/nodes/code/CodeNode.js:2]: I mistakenly thought for some reason that directory level of file has changed 🤦‍♂️

Comment (sunag) [reply examples/jsm/nodes/code/CodeNode.js:2]: No problems :)

File: examples/jsm/nodes/shadernode/ShaderNode.js [hunk 4]
```diff
@@ -255,6 +255,20 @@ const getAutoTypedConstNode = ( value ) => {
 
 };
 
+const getTypeFromInputNode = ( node ) => {
+
+	if ( node.isInputNode !== true || node.value === null ) return;
+
+	const value = node.value;
```
Comment (LeviPesin) [inline examples/jsm/nodes/shadernode/ShaderNode.js:262]: I think we should use `getValueType` function (from NodeUtils) here.

**Pull request merged**